### PR TITLE
Add support for named ports in autodiscover (hints & templates)

### PR DESCRIPTION
**Pull request by ChrsMark**
## What does this PR do?
Adds support for named ports in hints based autodiscover.
When for instance we have a target container like:
```
spec:
  containers:
    - name: natsqueue
      image: nats
      command: ["/nats-server"]
      args: ["-m", "8222"]
      ports:
        - name: web
          containerPort: 8222
          protocol: TCP
```
then we can reference the name of the port with `co.elastic.metrics/hosts: '${data.host}:${data.ports.web}'` in the annotations.

## Why is it important?
To cover cases like https://github.com/elastic/beats/issues/15796.

### Hints based Autodiscover
1. Deploy Metricbeat on k8s with hints autodiscover enabled:
```
metricbeat.autodiscover:
  providers:
    - type: kubernetes
      node: ${NODE_NAME}
      hints.enabled: true
```
2. Deploy a target Pod using the following spec:
```
apiVersion: v1
kind: Service
metadata:
  name: nats
  labels:
    app: nats
spec:
  clusterIP: None
  ports:
  - name: web
    port: 8222
    protocol: TCP
  selector:
    app: nats
  type: ClusterIP
---
apiVersion: v1
kind: Pod
metadata:
  name: nats
  labels:
    app: name
  annotations:
    co.elastic.metrics/module: nats
    co.elastic.metrics/metricsets: stats
    co.elastic.metrics/hosts: '${data.host}:${data.ports.web}'
spec:
  containers:
    - name: natsqueue
      image: nats
      command: ["/nats-server"]
      args: ["-m", "8222"]
      ports:
        - name: web
          containerPort: 8222
          protocol: TCP
```
3. Make sure that Nats metrics are being collected
4. Verify that old behaviour is not broken by deploying target with the following 2 different annotations:
`co.elastic.metrics/hosts: '${data.host}:8222'` and `co.elastic.metrics/hosts: '${data.host}:${data.port}`

### Template based Autodiscover
1. Deploy Metricbeat on k8s with hints autodiscover enabled:
```
metricbeat.autodiscover:
  providers:
    - type: kubernetes
      host: "${NODE_NAME}"
      templates:
        - condition:
            contains:
              kubernetes.annotations.nats.io/scrape: "true"
          config:
            - module: nats
              period: 10s
              hosts: ["${data.host}:${data.ports.web}"]
```
2. Deploy a target Pod using the following spec:
```
apiVersion: v1
kind: Service
metadata:
  name: nats
  labels:
    app: nats
spec:
  clusterIP: None
  ports:
  - name: web
    port: 8222
    protocol: TCP
  selector:
    app: nats
  type: ClusterIP
---
apiVersion: v1
kind: Pod
metadata:
  name: nats
  labels:
    app: name
  annotations:
    nats.io/scrape: "true"
spec:
  containers:
    - name: natsqueue
      image: nats
      command: ["/nats-server"]
      args: ["-m", "8222"]
      ports:
        - name: web
          containerPort: 8222
          protocol: TCP
```
3. Make sure that Nats metrics are being collected
4. Verify that old behaviour is not broken by deploying target with the following 2 different annotations:
`co.elastic.metrics/hosts: '${data.host}:8222'` and `co.elastic.metrics/hosts: '${data.host}:${data.port}`

### CODE REVIEW

File: libbeat/common/kubernetes/k8skeystore/kubernetes_keystore.go [hunk 1]
```diff
@@ -101,6 +101,9 @@ func NewKubernetesSecretsKeystore(keystoreNamespace string, ks8client k8s.Interf
 func (k *KubernetesSecretsKeystore) Retrieve(key string) (*keystore.SecureString, error) {
 	// key = "kubernetes.somenamespace.somesecret.value"
 	tokens := strings.Split(key, ".")
+	if tokens[0] != "kubernetes" {
+		return nil, keystore.ErrKeyDoesntExists
+	}
```
Comment (jsoriano) [inline libbeat/common/kubernetes/k8skeystore/kubernetes_keystore.go:106]: Unrelated change (and it should check if `tokens` has any elements before trying to get the first one :slightly_smiling_face:)

File: metricbeat/autodiscover/builder/hints/metrics.go [hunk 2]
```diff
@@ -88,6 +88,11 @@ func (m *metricHints) CreateConfig(event bus.Event, options ...ucfg.Option) []*c
 		noPort = true
 	}
 
+	portName, _ := event["port_name"].(string)
+	if portName == "" {
+		portName = ""
```
Comment (jsoriano) [inline metricbeat/autodiscover/builder/hints/metrics.go:93]: Redundant?

File: metricbeat/autodiscover/builder/hints/metrics.go [hunk 3]
```diff
@@ -192,13 +197,22 @@ func (m *metricHints) getMetricSets(hints common.MapStr, module string) []string
 	return msets
 }
 
-func (m *metricHints) getHostsWithPort(hints common.MapStr, port int, noPort bool) ([]string, bool) {
+func (m *metricHints) getHostsWithPort(hints common.MapStr, port int, noPort bool, portName string) ([]string, bool) {
 	var result []string
 	thosts := builder.GetHintAsList(hints, m.Key, hosts)
 
 	// Only pick hosts that have ${data.port} or the port on current event. This will make
 	// sure that incorrect meta mapping doesn't happen
```
Comment (jsoriano) [inline metricbeat/autodiscover/builder/hints/metrics.go:205]: This comment needs an update.

File: metricbeat/autodiscover/builder/hints/metrics.go [hunk 4]
```diff
@@ -214,6 +228,16 @@ func (m *metricHints) getHostsWithPort(hints common.MapStr, port int, noPort boo
 	return result, true
 }
 
+// getPortName returns the port name defined in a host hint ie: for ${data.host}:%{data.ports.web} returns web
```
Comment (jsoriano) [inline metricbeat/autodiscover/builder/hints/metrics.go:231]: The host hint can be an URL like `http://${data.host}:%{data.ports.prometheus}/metrics`, right? Would this method work on this case?
Could you add a test case for a case like this one?

File: metricbeat/autodiscover/builder/hints/metrics_test.go [hunk 5]
```diff
@@ -368,6 +368,33 @@ func TestGenerateHints(t *testing.T) {
 				},
 			},
 		},
+		{
+			message: "Named port in the hints should return the corresponding container port",
+			event: bus.Event{
+				"host":      "1.2.3.4",
+				"port":      3306,
+				"port_name": "some",
```
Comment (jsoriano) [inline metricbeat/autodiscover/builder/hints/metrics_test.go:376]: If `port_name` is going to be included in all events we should probably add a `port_name` to the rest of test cases.
Or at least add a test case with an event with `port_name`, but with a hosts hint that doesn't use named ports.

File: metricbeat/autodiscover/builder/hints/metrics.go (hunk 4 above)
Comment (ChrsMark) [reply metricbeat/autodiscover/builder/hints/metrics.go]: That needed some tuning! Thanks for mentioning this case!

**Comment by ChrsMark** [comment]
@jsoriano thanks for the review. Comments have been addressed, let me know what you think!

File: metricbeat/autodiscover/builder/hints/metrics.go [hunk 6]
```diff
@@ -228,8 +228,16 @@ func (m *metricHints) getHostsWithPort(hints common.MapStr, port int, noPort boo
 	return result, true
 }
 
-// getPortName returns the port name defined in a host hint ie: for ${data.host}:%{data.ports.web} returns web
+// getPortName returns the port name defined in a host hint ie: for ${data.host}:${data.ports.web} returns web
 func (m *metricHints) getPortName(h string) string {
+
+	re := regexp.MustCompile(`\${data.ports.*\}`)
```
Comment (jsoriano) [inline metricbeat/autodiscover/builder/hints/metrics.go:234]: Nit. This regexp could be compiled to a package variable so it doesn't need to be re-compiled for each port.

File: libbeat/common/kubernetes/k8skeystore/kubernetes_keystore.go [hunk 7]
```diff
@@ -101,6 +101,9 @@ func NewKubernetesSecretsKeystore(keystoreNamespace string, ks8client k8s.Interf
 func (k *KubernetesSecretsKeystore) Retrieve(key string) (*keystore.SecureString, error) {
 	// key = "kubernetes.somenamespace.somesecret.value"
 	tokens := strings.Split(key, ".")
+	if len(tokens) > 0 && tokens[0] != "kubernetes" {
+		return nil, keystore.ErrKeyDoesntExists
+	}
```
Comment (jsoriano) [inline libbeat/common/kubernetes/k8skeystore/kubernetes_keystore.go:106]: Does this fix need a changelog entry? Should it be done in a different PR?

File: metricbeat/autodiscover/builder/hints/metrics_test.go [hunk 8]
```diff
@@ -288,6 +288,7 @@ func TestGenerateHints(t *testing.T) {
 			event: bus.Event{
 				"host": "1.2.3.4",
 				"port": 0,
+				"port_name": "prometheus",
```
Comment (jsoriano) [inline metricbeat/autodiscover/builder/hints/metrics_test.go:291]: Nit. This event doesn't have a port, could it have a `port_name`?
```suggestion
```

File: metricbeat/autodiscover/builder/hints/metrics.go [hunk 9]
```diff
@@ -214,6 +228,24 @@ func (m *metricHints) getHostsWithPort(hints common.MapStr, port int, noPort boo
 	return result, true
 }
 
+// getPortName returns the port name defined in a host hint ie: for ${data.host}:${data.ports.web} returns web
+func (m *metricHints) getPortName(h string) string {
+
```
Comment (jsoriano) [inline metricbeat/autodiscover/builder/hints/metrics.go:233]: Nit.
```suggestion
```

File: libbeat/common/kubernetes/k8skeystore/kubernetes_keystore.go (hunk 7 above)
Comment (ChrsMark) [reply libbeat/common/kubernetes/k8skeystore/kubernetes_keystore.go:106]: This one is mostly to "mute" some DEBUG messages that are shown in the console when configs are being evaluated at https://github.com/elastic/beats/blob/e523380e7b51bc482f78428c7c54de0de25dc2e4/libbeat/autodiscover/template/config.go#L135.

With this addition we make sure that we don't log DEBUG msgs for keys that come from unrelated configs (other than `kubernetes.*`)

**Comment by exekias** [comment]
Sorry I'm late to this one but have some comments:

I see most of the logic here is done in the hints system, while I would expect this info (port names) to be even more useful in static templates.

If I'm configuring a Pod I already know the ports I'm exposing, so I can easily use the port number in the hints. On the other side, having this level of indirection helps me in the templates, as I (monitoring system maintainer) can tell whoever is deploying Pods (developers) to use certain port name to get some behaviour (ie, expose HTTP metrics under a port called http).

Would it make sense to just expose a `ports` map in the autodiscover Pod event, instead of specific data in container events?

**Comment by ChrsMark** [comment]
It seems like a +1 by for the static configs' addition.

How about having both? Having both will give a more unified of configuration for both approaches (I ve seen the pain of having part of functionalities supported on static configs but not in hints with Keystore cases and I think we could avoid having an extra casing scenario here).

So wdyt @exekias about keeping this one and come with a follow up PR with the respective part for static configs?

**Comment by exekias** [comment]
SGTM! I understand that variable names will be the same, so from the user POV the experience is unified

**Comment by ChrsMark** [comment]
This PR was updated in order to cover not only hints based Autodiscover but also  static Autodiscover templates.

@exekias @jsoriano could you please have another look?

File: metricbeat/docs/autodiscover-kubernetes-config.asciidoc [hunk 10]
```diff
@@ -34,6 +34,9 @@ Pods share an identical host. If only the `{data.host}` variable is interpolated
 then one config will be generated per host. The configs will be identical.
 After they are de-duplicated, only one will be used.
 
+In order to target one specific exposed port `{data.host}:{data.port.web}` can be used
+in template config, where `web` is the name of the exposed container port.
+
```
Comment (jsoriano) [inline metricbeat/docs/autodiscover-kubernetes-config.asciidoc:39]: Can we add some test for this? It could be done in a follow up if you prefer, I couldn't find tests for `data.port` so we may need to add more things.

File: metricbeat/docs/autodiscover-kubernetes-config.asciidoc [hunk 11]
```diff
@@ -34,6 +34,9 @@ Pods share an identical host. If only the `{data.host}` variable is interpolated
 then one config will be generated per host. The configs will be identical.
 After they are de-duplicated, only one will be used.
 
+In order to target one specific exposed port `{data.host}:{data.port.web}` can be used
```
Comment (jsoriano) [inline metricbeat/docs/autodiscover-kubernetes-config.asciidoc:37]:
```suggestion
In order to target one specific exposed port `{data.host}:{data.ports.web}` can be used
```

**Comment by ChrsMark** [comment]
Thanks for the reviews @kaiyan-sheng and @jsoriano!

I've added a couple of tests for config templates too.

File: libbeat/autodiscover/template/config_test.go [hunk 12]
```diff
@@ -74,6 +79,38 @@ func TestConfigsMapping(t *testing.T) {
...
+    hosts: ["${data.host}:${data.ports.web}"]`,
+			event: bus.Event{
+				"foo":  3,
+				"host": "1.2.3.4",
+				"ports": common.MapStr{
+					"web": 8080,
+				},
+			},
+			expected: []*common.Config{configPorts},
+		},
```
Comment (jsoriano) [inline libbeat/autodiscover/template/config_test.go:98]: :+1:

**Comment by ChrsMark** [comment]
Close  #15796
